tests/msc2716_test.go

```diff
@@ -0,0 +1,643 @@
+		newEvent := b.Event{
+			Type: "m.room.message",
+			Content: map[string]interface{}{
+				"msgtype": "m.text",
+				"body":    fmt.Sprintf("Message %d", i),
+			},
+		}
+		newEventId := c.SendEventSynced(t, roomID, newEvent)
+		evs[i] = newEventId
+	}
+
+	return evs
+}
+
+var chunkCount int64 = 0
+
+func backfillBatchHistoricalMessages(
+	t *testing.T,
+	c *client.CSAPI,
+	virtualUserID string,
+	roomID string,
+	insertAfterEventId string,
+	insertTime time.Time,
+	chunkID string,
+	count int,
+	status int,
+) (res *http.Response) {
+	// Timestamp in milliseconds
+	insertOriginServerTs := uint64(insertTime.UnixNano() / int64(time.Millisecond))
+
+	timeBetweenMessagesMS := uint64(TimeBetweenMessages / time.Millisecond)
+
+	evs := make([]map[string]interface{}, count)
+	for i := 0; i < len(evs); i++ {
+		newEvent := map[string]interface{}{
+			"type":             "m.room.message",
+			"sender":           virtualUserID,
+			"origin_server_ts": insertOriginServerTs + (timeBetweenMessagesMS * uint64(i)),
+			"content": map[string]interface{}{
+				"msgtype":          "m.text",
+				"body":             fmt.Sprintf("Historical %d (chunk=%d)", i, chunkCount),
+				MSC2716_HISTORICAL: true,
+			},
+		}
+
+		evs[i] = newEvent
+	}
+
+	joinEvent := map[string]interface{}{
+		"type":             "m.room.member",
+		"sender":           virtualUserID,
+		"origin_server_ts": insertOriginServerTs,
+		"content": map[string]interface{}{
+			"membership": "join",
+		},
+		"state_key": virtualUserID,
+	}
+
+	query := make(url.Values, 2)
+	query.Add("prev_event", insertAfterEventId)
+	query.Add("user_id", virtualUserID)
+	// If provided, connect the chunk to the last insertion point
+	if chunkID != "" {
+		query.Add("chunk_id", chunkID)
+	}
+
+	b, err := json.Marshal(map[string]interface{}{
+		"events":                evs,
+		"state_events_at_start": []map[string]interface{}{joinEvent},
+	})
+	if err != nil {
+		t.Fatalf("msc2716.backfillBatchHistoricalMessages failed to marshal JSON body: %s", err)
+	}
+
+	res = c.DoFunc(
+		t,
+		"POST",
+		[]string{"_matrix", "client", "unstable", "org.matrix.msc2716", "rooms", roomID, "batch_send"},
+		client.WithRawBody(b),
+		client.WithContentType("application/json"),
+		client.WithQueries(query),
+	)
+	// Save the body so we can re-create after the buffer closes
+	body := client.ParseJSON(t, res)
+	// Since the original body can only be read once, create a new one from the body bytes we just read
+	res.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+	must.MatchResponse(t, res, match.HTTPResponse{
+		StatusCode: status,
+	})
+	// After using up the body in the must.MatchResponse above, create the body again
+	// Since the original body can only be read once, create a new one from the body bytes we just read
+	res.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+
+	chunkCount++
+
+	return res
+}
+
+func getEventsFromBatchSendResponse(t *testing.T, res *http.Response) (stateEventsIDs []string, eventIDs []string) {
```

Comment:
All usages of this function immediately `_` the `stateEventIDs`. Either you aren't testing something you should be, or this is spurious and should be removed.

Reply:
:+1: Removed